Add a get IRC27NFT Metadata guide (#1583)
iotaledger/iota-wiki: docs/build/isc/v1.0.0-rc.6/docs/how-tos/core-contracts/nft/get-nft-metadata.md

# Get NFT Metadata

This guide explains how to use the [`getIRC27NFTData`](../../../reference/magic-contract/ISCSandbox.md#getirc27nftdata) function within a smart contract to fetch information about a specific IRC27 NFT on the IOTA Network.

<GetNftMetadata />

## Understanding the `getIRC27NFTData` Function

The `getIRC27NFTData` function retrieves metadata for an IRC27 NFT based on its identifier. IRC27 is a series of standards to support interoperable and universal NFT systems throughout the IOTA ecosystem.

### Function Signature:

```solidity
function getIRC27NFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData);
```

### Parameters:

`NFTID id`: This parameter represents the unique identifier of the IRC27 NFT you intend to retrieve information about.

### Return Value:

The function returns a data structure of type IRC27NFT. This structure incorporates two elements:
*  `nft`: This element provides details regarding the underlying on-chain NFT. It's of type `ISCNFT`.
*  `metadata`: This element offers information specific to the IRC27 standard. It's of type `IRC27NFTMetadata`.

> 

Author: Thanks, deleted as suggested.

import "@iota/iscmagic/ISC.sol";
import "@iota/iscmagic/ISCTypes.sol";

contract NFTMetadata {

function fetchNFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData) {
        require(nftID.length == 32, "NFT ID must be 32 bytes");
        bytes32 id = bytes32(nftID);
        NFTID nftIDTyped = NFTID.wrap(id);
        irc27NftData = ISC.sandbox.getIRC27NFTData(nftIDTyped);
        return irc27NftData;
    }

function encodeNFTMetadata(
    string memory _standard,
    string memory _version,
    string memory _mimeType,
    string memory _uri,
    string memory _name,
    string memory _description
) 
    public pure returns (IRC27NFTMetadata memory) 
{
    string memory json = string(abi.encodePacked(
        '{"name":"', _name,
        '","description":"', _description,
        '","image":"', _uri,
        '"}'
    ));

    string memory encodedURI = string(abi.encodePacked(
        "data:application/json;base64,", 
        Base64.encode(bytes(json))
    ));

    return IRC27NFTMetadata({
        standard: _standard,
        version: _version,
        mimeType: _mimeType,
        uri: encodedURI,
        name: _name
    });
}
}
```
## Example Usage

> To mint a new NFT and fetch its metadata, you would use the above functions within a larger contract that handles NFT minting and management.

Author: Ok thanks, updated the doc

---
description: How to get NFT metadata on L1
image: /img/logo/WASP_logo_dark.png
tags:
  - NFT
  - EVM
  - how-to
---
import GetNftMetadata from '../../../_partials/how-tos/token/_get-nft-metadata.md';

# Get NFT Metadata

> This guide explains how to utilize the `getIRC27NFTData` function within a smart contract to fetch information about a specific IRC27 NFT on the IOTA Network. The function resides in the ISCSandbox contract and returns details pertaining to the requested NFT.

Reviewer: ```suggestion
This guide explains how to use the [`getIRC27NFTData`](../../../reference/magic-contract/ISCSandbox/#getirc27nftdata) function within a smart contract to fetch information about a specific IRC27 NFT on the IOTA Network. The function resides in the ISCSandbox contract and returns details pertaining to the requested NFT.
```

We could also cut the last sentence if we add the link to the refs.

Author: I have made the necessary updates based on your suggestions and also removed the last sentence as requested.

Author: getIRC27NFTData url was broken, fixed it with this url: https://github.com/iotaledger/wasp/blob/develop/packages/vm/core/evm/iscmagic/ISCSandbox.sol#L118

Reviewer 2: I will fix it. But we need the updated reference docs first 👍

> 

Reviewer: ```suggestion

<GetNftMetadata />

```

## Understanding the `getIRC27NFTData` Function

The `getIRC27NFTData` function retrieves metadata for an IRC27 NFT based on its identifier. IRC27 is a series of standards to support interoperable and universal NFT systems throughout the IOTA ecosystem.

> ## Function Signature:
> ```solidity
> function getIRC27NFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData);
> ```
> ## Parameters:
> `NFTID id`: This parameter represents the unique identifier of the IRC27 NFT you intend to retrieve information about.
> ## Return Value:
> The function returns a data structure of type IRC27NFT. This structure incorporates two elements:
> *  `nft`: This element provides details regarding the underlying on-chain NFT. It's of type `ISCNFT`.
> *  `metadata`: This element offers information specific to the IRC27 standard. It's of type `IRC27NFTMetadata`.

Reviewer: ````suggestion
### Function Signature:

```solidity
function getIRC27NFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData);
```

### Parameters:

`NFTID id`: This parameter represents the unique identifier of the IRC27 NFT you intend to retrieve information about.

### Return Value:

The function returns a data structure of type IRC27NFT. This structure incorporates two elements:
*  `nft`: This element provides details regarding the underlying on-chain NFT. It's of type `ISCNFT`.
*  `metadata`: This element offers information specific to the IRC27 standard. It's of type `IRC27NFTMetadata`.
````

This probably belongs in the refs, though I think they are autogenerated.

Author: I made the changes, but like you said it belongs in the refs, let me know if I should take this section out

> ### Here's a step-by-step breakdown of how to leverage getIRC27NFTData to acquire IRC27 NFT metadata:

Reviewer: ```suggestion
## How To Use `getIRC27NFTData` 
```

Author: Thanks, done

> 1. Fetching IRC27 NFT Metadata:

Reviewer: ```suggestion
### 1. Fetch IRC27 NFT Metadata
```

Author: Done

> Create a function called fetchNFTData in your contract that calls getIRC27NFTData and processes its return value.

Reviewer: ```suggestion
Create a function called `fetchNFTData` in your contract that calls `getIRC27NFTData` and processes its return value.
```
We usually wrap function names in ``

Author: Thanks, noted.

```solidity

> // Function to get NFT data
>     function fetchNFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData) {
>         bytes32 id = bytes32(nftID);
>         NFTID nftIDTyped = NFTID.wrap(id);
>         irc27NftData = ISC.sandbox.getIRC27NFTData(nftIDTyped);
>         return irc27NftData;
>     }

Reviewer: ```suggestion
function fetchNFTData(bytes memory nftID) public view returns (IRC27NFT memory irc27NftData) {
    bytes32 id = bytes32(nftID);
    NFTID nftIDTyped = NFTID.wrap(id);
    irc27NftData = ISC.sandbox.getIRC27NFTData(nftIDTyped);
    return irc27NftData;
}
```

Author: Done

```

> 2. Encoding URI Data in IRC27NFTMetadata

Reviewer: ```suggestion
### 2. Encode the URI Data in `IRC27NFTMetadata`
```

Author: Done

The `IRC27NFTMetadata` struct holds detailed metadata, including the URI and other fields like name and description. Encoding this data correctly is crucial for ensuring the integrity and accessibility of the NFT's metadata.

```solidity
struct IRC27NFTMetadata {string uri; string name;string description;}
```

> 3. Implementing URI Encoding

Reviewer: ```suggestion
### 3. Implement URI Encoding
```

Author: Thanks, it's updated

> The OpenZeppelin Contracts library provides a `Base64` utility that simplifies encoding URI data. Here's how to implement URI encoding:

Reviewer: ```suggestion
The [OpenZeppelin Contracts](https://www.openzeppelin.com/) library provides a `Base64` utility that simplifies encoding URI data. Here's how to implement URI encoding:
```

Author: Thanks, updated

> ## Implementation:
> 

Reviewer: ```suggestion
```

Author: Done

```solidity
function encodeNFTMetadata(string memory _standard, string memory _version, 
    string memory _mimeType,
    string memory _uri,
    string memory _name,
    string memory _description
) public pure returns (IRC27NFTMetadata memory) 
{
    // Create JSON representation
    string memory json = string(abi.encodePacked(
        '{"name":"', _name,
        '","description":"', _description,
        '","image":"', _uri,
        '"}'
    ));

    // Base64 encode the JSON string
    string memory encodedURI = string(abi.encodePacked(
        "data:application/json;base64,", 
        Base64.encode(bytes(json))
    ));

    // Return the metadata struct
    return IRC27NFTMetadata({
        standard: _standard,
        version: _version,
        mimeType: _mimeType,
        uri: encodedURI,
        name: _name,
        description: _description
    });
}
```

> ## Full Example Contract
> Combining all the above steps, here’s a complete example:

Reviewer: ```suggestion
## Full Example Contract

Combining all the above steps, here’s a complete example:
```

        bytes32 id = bytes32(nftID);
        NFTID nftIDTyped = NFTID.wrap(id);
        irc27NftData = ISC.sandbox.getIRC27NFTData(nftIDTyped);
        return irc27NftData;
    }

function encodeNFTMetadata(
    string memory _standard,
    string memory _version,
    string memory _mimeType,
    string memory _uri,
    string memory _name,
    string memory _description
) 
    public pure returns (IRC27NFTMetadata memory) 
{
    // Create JSON representation
    string memory json = string(abi.encodePacked(
        '{"name":"', _name,
        '","description":"', _description,
        '","image":"', _uri,
        '"}'
    ));

    // Base64 encode the JSON string
    string memory encodedURI = string(abi.encodePacked(
        "data:application/json;base64,", 
        Base64.encode(bytes(json))
    ));

    // Return the metadata struct
    return IRC27NFTMetadata({
        standard: _standard,
        version: _version,
        mimeType: _mimeType,
        uri: encodedURI,
        name: _name
    });
}
}
```
## Example Usage

To mint a new NFT and fetch its metadata, you would use the above functions within a larger contract that handles NFT minting and management.

> <GetNftMetadata />

Reviewer: ```suggestion
```

Author: Done.
